### PR TITLE
feature: add pouch rmi command

### PR DESCRIPTION
**Pull request by HusterWan**
Signed-off-by: HusterWan <zirenwan@gmail.com>

**1.Describe what this PR did**

add cli pouch rmi command

**2.Does this pull request fix one issue?** 
fixes https://github.com/alibaba/pouch/issues/268

**3.Describe how you did it**

**4.Describe how to verify it**
```
[root@docker-registry github.com]# pouch rmi docker.io/library/nginx:latest
docker.io/library/nginx:latest
```
**5.Special notes for reviews**

- [x] delete image by image name or ID

- [ ] check if image is being used by any container when deleting

- [x] define api in swagger.yaml 

- [x] add pouch rmi command test unit

### CODE REVIEW

File: cli/rmi.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	baseCommand
+	force bool
+}
+
+// Init initialize rmi command
+func (cc *RmiCommand) Init(c *Cli) {
+	cc.cli = c
+	cc.cmd = &cobra.Command{
+		Use:   "rmi image ",
+		Short: "Remove one or more images by reference",
```
Comment (allencloud) [inline cli/rmi.go]: Please add Long and Example here like create command https://github.com/alibaba/pouch/blob/master/cli/create.go#L25-L34:
```
cc.cmd = &cobra.Command{
		Use:   "create [image]",
		Short: "Create a new container with specified image",
		Long:  createDescription,
		Args:  cobra.MinimumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return cc.runCreate(args)
		},
		Example: createExample(),
	}
```
These things are merged this morning and useful to auto generate markdown document, so you may have no idea about that now.

File: cli/rmi.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+		Args:  cobra.MinimumNArgs(1),
```
Comment (allencloud) [inline cli/rmi.go]: s/cobra.MinimumNArgs(1)/cobra.ExactArgs(1)/

File: apis/server/image_bridge.go [hunk 3]
```diff
@@ -62,3 +64,17 @@ func (s *Server) searchImages(ctx context.Context, resp http.ResponseWriter, req
 	}
 	return json.NewEncoder(resp).Encode(response)
 }
+
+// removeImages delete image by reference
```
Comment (allencloud) [inline apis/server/image_bridge.go]: deletes images

File: client/image.go [hunk 4]
```diff
@@ -35,3 +38,28 @@ func (client *APIClient) ImageList() ([]types.Image, error) {
 	return imageList, err
 
 }
+
+// ImageRemoves removes an image
```
Comment (allencloud) [inline client/image.go]: Please add . at the end of this line, since godoc needs repo to do so.

File: daemon/mgr/image.go [hunk 5]
```diff
@@ -21,6 +21,9 @@ type ImageMgr interface {
 
 	// Search Images from specified registry.
 	SearchImages(ctx context.Context, name string, registry string) ([]types.SearchResultItem, error)
+
+	// RemoveImages remove an image by reference in local repo
```
Comment (allencloud) [inline daemon/mgr/image.go]: s/remove/removes

File: daemon/mgr/image.go [hunk 6]
```diff
@@ -89,3 +92,14 @@ func (mgr *ImageManager) SearchImages(ctx context.Context, name string, registry
...
 }
+
+// ImageRemoveOption wraps the image remove interface params.
+type ImageRemoveOption struct {
+	Force bool
+}
+
+// RemoveImages delete image by reference
+func (mgr *ImageManager) RemoveImages(ctx context.Context, name string, option *ImageRemoveOption) error {
+	err := mgr.client.RemoveImages(ctx, name)
```
Comment (allencloud) [inline daemon/mgr/image.go]: If that, just `return mgr.client.RemoveImages(ctx, name)`.

File: daemon/mgr/image.go [hunk 7]
```diff
@@ -89,3 +92,14 @@ func (mgr *ImageManager) SearchImages(ctx context.Context, name string, registry
 
 	return nil, nil
 }
+
+// ImageRemoveOption wraps the image remove interface params.
+type ImageRemoveOption struct {
+	Force bool
+}
+
+// RemoveImages delete image by reference
```
Comment (allencloud) [inline daemon/mgr/image.go]: deletes images

**Comment by HusterWan** [comment]
@allencloud，sorry push a not checked code, i will fix soon

File: ctrd/image.go [hunk 8]
```diff
@@ -21,8 +21,8 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-// RemoveImage removes an image.
-func (c *Client) RemoveImage(ctx context.Context, ref string) error {
+// RemoveImages deletes an image.
+func (c *Client) RemoveImages(ctx context.Context, ref string) error {
```
Comment (skoo87) [inline ctrd/image.go]: it's remove only one image, why name it RemoveImages ???

File: daemon/mgr/image.go [hunk 9]
```diff
@@ -89,3 +92,14 @@ func (mgr *ImageManager) SearchImages(ctx context.Context, name string, registry
 
 	return nil, nil
 }
+
+// ImageRemoveOption wraps the image remove interface params.
+type ImageRemoveOption struct {
+	Force bool
+}
```
Comment (skoo87) [inline daemon/mgr/image.go]: I think we can create the "image_types.go" file, put the "ImageRemoveOption" into it. 
WDYT ????

File: daemon/mgr/image.go [hunk 10]
```diff
@@ -89,3 +92,14 @@ func (mgr *ImageManager) SearchImages(ctx context.Context, name string, registry
...
+
+// RemoveImages deletes images by reference
+// TODO: Should support delete images by specify image ID
+func (mgr *ImageManager) RemoveImages(ctx context.Context, name string, option *ImageRemoveOption) error {
```
Comment (skoo87) [inline daemon/mgr/image.go]: why RemoveImages, not RemoveImage???

**Comment by skoo87** [comment]
In addition, there is API not found in swagger.yml.

**Comment by HusterWan** [comment]
@allencloud @skoo87 Thanks for your code review, code has updated, Please take another look

File: client/image.go [hunk 11]
```diff
@@ -35,3 +38,28 @@ func (client *APIClient) ImageList() ([]types.Image, error) {
...
+	}
+
+	resp, err := client.delete("/images/"+name, q)
+	if err != nil {
+		return err
+	}
+
+	defer ensureCloseReader(resp)
+
+	if resp.StatusCode != http.StatusNoContent {
```
Comment (allencloud) [inline client/image.go]: I think you defined a 2xx code to be 200 in swagger, but here you are comparing this with 204. Seems a little bit improper.

File: client/image.go [hunk 12]
```diff
@@ -35,3 +38,28 @@ func (client *APIClient) ImageList() ([]types.Image, error) {
...
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		return errors.New(string(body))
```
Comment (allencloud) [inline client/image.go]: if string(body) is empty string, this function will return an error. Is it reasonable.
Actually if  `resp.StatusCode != http.StatusNoContent` returns true, it means this is a non-2xx response. Then the error message will not be in `resp`, I think it is in `err` of `resp, err := client.delete("/images/"+name, q)`. This is guaranteed by HTTP framework. So I am wondering if this block is reasonable. 
Also we found similar code in other function of api client. Yes, if you agree with me, we should reconsider that part as well. @HusterWan @skoo87

Comment (HusterWan) [reply client/image.go]: @allencloud you are right, maybe we can return err directly here
``` 
_, err := client.delete("/images/"+name, q)

return err
```

File: test/pouch_cli_rmi_test.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+	// TODO add cleanup work
+}
+
+// TestRmiWorks tests "pouch rmi" work.
+func (suite *PouchRmiSuite) TestRmiWorks(c *check.C) {
+
+	// TODO: add wrong args.
+
+	cmd := exec.Command("pouch", "rmi", testImage)
+	runCmdPos(c, cmd)
```
Comment (allencloud) [inline test/pouch_cli_rmi_test.go]: I think it is proper to check `runCmdPos`. And it only checks whether the command executes well. I think we still need to checkout output to see if `testImage` is still in output. WDYT? @HusterWan @Letty5411

**Comment by allencloud** [comment]
LGTM, except one more check done in integration test.

File: apis/server/image_bridge.go [hunk 14]
```diff
@@ -62,3 +64,17 @@ func (s *Server) searchImages(ctx context.Context, resp http.ResponseWriter, req
 	}
 	return json.NewEncoder(resp).Encode(response)
 }
+
+// removeImage deletes an image by reference
+func (s *Server) removeImage(ctx context.Context, resp http.ResponseWriter, req *http.Request) error {
+	name := mux.Vars(req)["name"]
+
+	option := &ctrd.ImageRemoveOption{}
```
Comment (skoo87) [inline apis/server/image_bridge.go]: I think the "ImageRemoveOption" should be defined in mgr, not in ctrd.  mgr is the core control layer, the image_bridge layer can't not depend on ctrd directly.

File: daemon/mgr/image.go [hunk 15]
```diff
@@ -125,6 +128,16 @@ func (mgr *ImageManager) GetImage(ctx context.Context, idOrRef string) (*types.I
 	return mgr.cache.get(idOrRef)
 }
 
+// RemoveImage deletes an image by reference.
+func (mgr *ImageManager) RemoveImage(ctx context.Context, name string, option *ImageRemoveOption) error {
```
Comment (skoo87) [inline daemon/mgr/image.go]: We should remove the image from cache too .

File: test/pouch_cli_rmi_test.go [hunk 16]
```diff
@@ -0,0 +1,52 @@
+package main
+
+import (
+	"github.com/go-check/check"
```
Comment (skoo87) [inline test/pouch_cli_rmi_test.go]: modify import format ???

File: daemon/mgr/image.go (hunk 15 above)
Comment (HusterWan) [reply daemon/mgr/image.go]: you are right, i will fix it

**Comment by skoo87** [comment]
LGTM

**Comment by HusterWan** [comment]
image

@allencloud @skoo87 thanks a lot,  but i have not completed this problem yet,  we should check if image is being using by container. 
today is too busy to do it,  i will create a 'PR' to fix it this weekend.

**Comment by allencloud** [comment]
Thanks a lot. @HusterWan Actually our users are very urgent to experience this new feature. So we decided to merge this first. Yeah, we know that there is still one step missing. Maybe in a follow-up PR. Thanks again.

**Pull request merged**